### ISSUE TITLE
ResNet weight freeze

### ISSUE THREAD

**Issue by xingyul** [ppwwyyxx/tensorpack#296]
I played with ResNet-101 pre-trained model. I used 0 learning rate and 0 L2-regularization. I also used `tf.stop_gradient` as follows:

```
logits = (LinearWrap(image)
                      .Conv2D('conv0', 64, 7, stride=2, nl=BNReLU)
                      .MaxPooling('pool0', shape=3, stride=2, padding='SAME')
                      .apply(layer, 'group0', block_func, 64, defs[0], 1, first=True)
                      .apply(layer, 'group1', block_func, 128, defs[1], 2)
                      .apply(layer, 'group2', block_func, 256, defs[2], 2)
                      .apply(layer, 'group3', block_func, 512, defs[3], 2)
                      .BNReLU('bnlast')
                      .GlobalAvgPooling('gap')
                      .FullyConnected('linear', 1000, nl=tf.identity)())
logits = tf.stop_gradient(logits)
s = tf.get_variable('scale', shape=[1], dtype=tf.float32, initializer=tf.constant_initializer(1.0))
logits = s + logits
```
The above `s` variable is just used to ensure there's a variable trainable in the model. It doesn't change softmax result.
However, I found that after a few iterations of training, validation error changes. But it shouldn't change, since there's no gradient and 0 learning rate. `s` also shouldn't change prediction result.
Does tensorpack include something magical that causes this happening?

**Comment by ppwwyyxx** [comment]
batch norm uses moving averages which get changed whenever training data pass through.

**Comment by ppwwyyxx** [comment]
You can disable moving average update by overwrite the default callback list (`extra_callbacks`). http://tensorpack.readthedocs.io/en/latest/tutorial/callback.html

**Comment by xingyul** [comment]
Thanks a lot! It works.

**Comment by xingyul** [comment]
Is there a way to freeze only part of batchnorm layers? My project requires freezing part of the model and fine-tuning the rest of it. Thanks.

**Comment by ppwwyyxx** [comment]
https://github.com/ppwwyyxx/tensorpack/issues/87#issuecomment-270545291

**Comment by xingyul** [comment]
I think batch normalization layer is updated through moving averages. So even if gradient is zero, batch normalization parameters will still change. Is there a way to only apply `RunUpdateOps()` to some of batch normalzation layers?

**Comment by ppwwyyxx** [comment]
I think bn averages should be update when you fine tune.

If you'd like to only update some of your bn ops, you can write a new callback similar to `RunUpdateOps`, and filter the ops by your self. The code is very simple:
```python
class RunUpdateOps(RunOp):
    def __init__(self, collection=tf.GraphKeys.UPDATE_OPS):
        def f():
            ops = tf.get_collection(collection)
# do your filtering here
            if ops:
                logger.info("Applying UPDATE_OPS collection of {} ops.".format(len(ops)))
                return tf.group(*ops, name='update_ops')
            else:
                return tf.no_op(name='empty_update_ops')

        super(RunUpdateOps, self).__init__(
            f, run_before=False, run_as_trigger=False, run_step=True)
```
